### ISSUE TITLE
When trying to activate the extension in pdf shared on Dropbox, pdf viewer fails

### ISSUE THREAD

**Issue by lenazun** [hypothesis/h#2685]
Steps to repro:
1. Receive a link to a pdf hosted on Dropbox (ie. https://www.dropbox.com/s/9oj8z7au1dlvb8e/felicidade.pdf?dl=0)
2. Click on the H icon to activate the extension

Expected results:
PDF viewer is opened. Sidebar is displayed (hopefully!) with existing annotations

Actual results:
PDF viewer is opened. User gets error:
Invalid or corrupted PDF file.
PDF.js v1.1.215 (build: c9a7498)
Message: Invalid PDF structure

**Comment by tilgovi** [comment]
The extension is detecting a PDF because the URL ends in .pdf but the resource at that URL is not a PDF, it's a Dropbox PDF viewer.

**Comment by tilgovi** [comment]
I can tell from the markup that they are using PDF.js, but it's in an iframe on their viewer page.

**Comment by judell** [comment]
https://www.dropbox.com/s/9oj8z7au1dlvb8e/felicidade.pdf?dl=1 (vs dl=0) will fetch the PDF instead of the viewer. But we still don't activate.

curl --head https://www.dropbox.com/s/9oj8z7au1dlvb8e/felicidade.pdf?dl=1 

```
HTTP/1.1 302 FOUND
Server: nginx
Date: Thu, 19 Nov 2015 18:25:36 GMT
Content-Type: text/html; charset=utf-8
Connection: keep-alive
x-xss-protection: 1; mode=block
x-robots-tag: noindex, nofollow, noimageindex
content-security-policy: sandbox
set-cookie: locale=en; Domain=dropbox.com; expires=Tue, 17 Nov 2020 18:25:36 GMT; Path=/; secure
set-cookie: gvc=OTAzNjY4NjIyNzYxMDYzNjkzNzY3NDk1NDg0NzAzNjQ0MjgzNQ%3D%3D; expires=Tue, 17 Nov 2020 18:25:36 GMT; httponly; Path=/; secure
set-cookie: t=HwYaBjB253LhX-Bd_Z_uS5h7; Domain=dropbox.com; expires=Sun, 18 Nov 2018 18:25:36 GMT; httponly; Path=/; secure
set-cookie: js_csrf=HwYaBjB253LhX-Bd_Z_uS5h7; Domain=www.dropbox.com; expires=Sun, 18 Nov 2018 18:25:36 GMT; Path=/; secure
x-dropbox-request-id: 5f4381ffb9e0288abd68184052c14555
pragma: no-cache
cache-control: no-cache
x-dropbox-http-protocol: None
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
location: https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1
Strict-Transport-Security: max-age=15552000; includeSubDomains
```

In other cases, this hack would force activation:

https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1&foo=bar.pdf

But not here.

@BigBlueHat what do you think?

**Comment by BigBlueHat** [comment]
@judell the response is a `302 Found` so it's redirecting to the value of the `Location` header:
```
...
location: https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1
```

Testing with `curl` breaks down though because Dropbox is using some sort of temporary "auth" to fingerprint the redirect.

So this request on its own doesn't work (for example).
```
curl --head https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1
```

However, my guess is that (if properly "authenticated") that request would return a `200 OK` with `Content-Type: application/pdf`--and the fix for that is up in PR https://github.com/hypothesis/h/pull/2437

**Comment by tilgovi** [comment]
> https://www.dropbox.com/s/9oj8z7au1dlvb8e/felicidade.pdf?dl=1 (vs dl=0) will fetch the PDF instead of the viewer. But we still don't activate.

Isn't that #1900?

**Comment by BigBlueHat** [comment]
@tilgovi yep. Fix in #2437 (but lacks unit tests).

**Comment by judell** [comment]
@BigBlueHat Yes I noted the redirect in https://github.com/hypothesis/h/issues/2685#issuecomment-158148353 and also noted there that https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1&foo=bar.pdf in this case does not fool the extension into loading as it does in other cases, for example our canonical one:

http://www.plosone.org/article/fetchObject.action?uri=info:doi/10.1371/journal.pone.0115253&representation=PDF

vs.

http://www.plosone.org/article/fetchObject.action?uri=info:doi/10.1371/journal.pone.0115253&representation=PDF&foo=bar.pdf

I'm curious how that differs from:

https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1

vs

https://dl.dropboxusercontent.com/content_link/3U1fSB2i9sapQgME5SY27rHrpqA5ySJkXhUax936UPlYiNYXJgz3gJUQzGJzbD56/file?dl=1&foo=bar.pdf

**Comment by BigBlueHat** [comment]
@judell my guess would be a `Content-Disposition` header's in there post-redirect. For me, the download is "forced"--it doesn't view it in Chrome's built in viewer. Same in FF.

**Comment by tilgovi** [comment]
As @BigBlueHat pointed out, the redirect sets some cookies for authorization. If you use `curl -L` on the original URL you will get the PDF content.

So, the hack would, if it were possible here, be this url:

https://www.dropbox.com/s/9oj8z7au1dlvb8e/felicidade.pdf?dl=1&foo=bar.pdf

But, I think @BigBlueHat is right. Our extension doesn't try to prevent you from downloading the PDF, so if the Content-Disposition is set for download of an attachment I don't think the viewer will show.

**Comment by judell** [comment]
Ah. Now it's coming back to me.

content-disposition: attachment; filename="felicidade.pdf";
filename*=UTF-8''felicidade.pdf

Right.

This is perhaps a thing to put into a yet-to-be-written guide for
publishers who wish to host H-compatible content.

Unless we think the extension could and should "try to prevent you from
downloading the PDF"?

**Comment by BigBlueHat** [comment]
@judell definitely not. Don't break the Web, please. :smile:

**Comment by tilgovi** [comment]
> Unless we think the extension could and should "try to prevent you from downloading the PDF"?

It could initially do so, and then offer a button in the UI to download. That gets into more extensive modification to the PDF.js viewer, more local state tracking, more opportunities for bugs.

**Comment by tilgovi** [comment]
There's a possibility that hasn't been mentioned here and that's to close this issue one of two ways.

- Won't fix. If Dropbox doesn't want to serve the PDF straight to the browser viewer, Dropbox is not an acceptable place for hosting PDFs if you want people to annotate them with Hypothesis.

- Duplicate of #2134

**Comment by BigBlueHat** [comment]
I'd mark as "won't fix" as the URL because...
 - `dl=0` is for a non-PDF URL (and, if annotating inside that iframe HTML page is wanted...then see #2134).
 - `dl=1` is not meant for "serving" the PDF, but expressly for downloading it. Changing that intended functionality would (imo) "break the Web"--i.e. if `dl=1` didn't result in a file on my hard drive (without further action), then clicking the "Download" button in the Dropbox UI (etc) would not do what I expected.

@judell I'll let you make the official call, but I'd say "won't fix" + reference this issue from #2134 (...as if it isn't already :wink:).

**Comment by judell** [comment]
Won't fix.
